### ISSUE TITLE
Pylance behaviour on overloads doesn't  match pyrights behaviour

### ISSUE THREAD

**Issue by YodaPY** [microsoft/pylance-release#6277]
<h3> Environment data </h3>

-   Language Server version: 2024.8.1
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: off

<h3> Code Snippet </h3>

```python
import typing as t
from typing_extensions import ParamSpec, Concatenate

_OriginArgsP = ParamSpec("_OriginArgsP")
_OriginSelfT = t.TypeVar("_OriginSelfT")
_OriginReturnT = t.TypeVar("_OriginReturnT")

_TargetSelfT = t.TypeVar("_TargetSelfT")
_TargetArgsP = ParamSpec("_TargetArgsP")
_NewReturnT = t.TypeVar("_NewReturnT")

@t.overload
def copy_method_signature(
    origin: t.Callable[Concatenate[_OriginSelfT, _OriginArgsP], _OriginReturnT], *, return_type: None=..., skip_self: t.Literal[False]=...
) -> t.Callable[[t.Callable[Concatenate[_TargetSelfT, _TargetArgsP], t.Any]], t.Callable[Concatenate[_TargetSelfT, _OriginArgsP], _OriginReturnT]]:
    """A method copies the full signature from another method"""

@t.overload
def copy_method_signature(
    origin: t.Callable[Concatenate[_OriginSelfT, _OriginArgsP], _OriginReturnT], *, return_type: type[_NewReturnT], skip_self: t.Literal[False]=...
) -> t.Callable[[t.Callable[Concatenate[_TargetSelfT, _TargetArgsP], t.Any]], t.Callable[Concatenate[_TargetSelfT, _OriginArgsP], _NewReturnT]]:
    """A method copies the parameter signature from another method but overrides the return type"""

@t.overload
def copy_method_signature(
    origin: t.Callable[Concatenate[_OriginSelfT, _OriginArgsP], _OriginReturnT], *, return_type: None=..., skip_self: t.Literal[True]
) -> t.Callable[[t.Callable[_TargetArgsP, t.Any]], t.Callable[_OriginArgsP, _OriginReturnT]]:
    """A function copies the full signature from another method"""

@t.overload
def copy_method_signature(
    origin: t.Callable[Concatenate[_OriginSelfT, _OriginArgsP], _OriginReturnT], *, return_type: type[_NewReturnT], skip_self: t.Literal[True]
) -> t.Callable[[t.Callable[_TargetArgsP, t.Any]], t.Callable[_OriginArgsP, _NewReturnT]]:
    """A function copies the paramater signature from another method but overrides the return type"""

@t.no_type_check
def copy_method_signature(
    origin, *, return_type=None, skip_self=False
):
    @t.no_type_check
    def inner(target):
        return target
    
    return inner

class A:
    @t.overload
    def method(self, a: int)  -> None:
        ...

    @t.overload
    def method(self, a: str) -> int:
        ...

    def method(self, a: t.Union[int, str]) -> t.Optional[int]:
        ...

@copy_method_signature(A.method, return_type=int, skip_self=True)
@t.no_type_check
def hello(*args, **kwargs):
    ...


reveal_type(hello)
```

<h3> Repro Steps </h3>

1. Hover over the `copy_method_signature` decorator

<h3> Expected behavior </h3>

No error

<h3> Actual behavior </h3>

`No overloads for "copy_method_signature" match the provided argumentsPylance[reportCallIssue](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportCallIssue)
method.py(31, 5): Overload 4 is the closest match`

2. Hover over the `reveal_type` call

<h3> Expected behaviour </h3>

`Overload[(a: int) -> int, (a: str) -> int]` as pointed out by  pyright in  the command  line:
`method.py:64:13 - information: Type of "hello" is "Overload[(a: int) -> int, (a: str) -> int]"
0 errors, 0 warnings, 1 information `

<h3> Actual behaviour </h3>

`Type of "hello" is "(...) -> None"`

<h3> Logs </h3>

```
XXX
```

**Comment by YodaPY** [comment]
This only seems to happen  if you pass in arguments to the  `copy_method_signature` call
```python
@copy_method_signature(A.method)
@t.no_type_check
def hello(*args, **kwargs):
    ...
``` 
pylance seems to  have no problem with this

**Comment by bschnurr** [comment]
thanks for the report. pylance didn't get update last week and I think pyright recently updated its overload logic, so I'm guessing that is the reason for the disagreement. In the mean time you can check which version of pyright pylance is using in the python language server output window.

**Comment by YodaPY** [comment]
Yup,  pylance is using version 1.1.373, seems like this bug was fixed in pyright version 1.1.374

**Comment by debonte** [comment]
> Yup, pylance is using version 1.1.373, seems like this bug was fixed in pyright version 1.1.374

Pylance's latest prerelease build (2024.8.101) is using Pyright 1.1.375.

**Comment by YodaPY** [comment]
Ah yep, can confirm it's working on the prerelease build, thanks a lot!
